Blogging Prompts: Migrate to v3 logic

dvdchr opened the pull request:
Refs https://github.com/wordpress-mobile/WordPress-iOS/issues/20496

> **Note**
> Originally, this was meant to be done partially in WordPressKit (see https://github.com/wordpress-mobile/WordPressKit-iOS/pull/592), but I've decided to port the logic here since the Jetpack app is only the sole consumer of Blogging Prompts. 
> 
> See https://github.com/wordpress-mobile/WordPressKit-iOS/pull/592#issuecomment-1788042166 for more context. 

This PR migrates the Blogging Prompts fetching mechanism to the v3 version. Just so you know, the new endpoint version **will return a different set of prompts** when compared to v2, despite having the same date range. Some other notes:

- With the fetching logic now ported to `WordPress-iOS`, we now use `BloggingPromptRemoteObject` to represent a blogging prompt object from the v3 API.
  - I didn't port the Blogging Prompt Settings logic just yet to keep things manageable. It's a low priority, but once it's ported, we can remove the dependency on `BloggingPromptServiceRemote`.
- The `upsert` method is updated to overwrite based on `date` instead of `promptID`. Furthermore, the method ensures that there will only be one prompt per date. When processing new prompts, the first matching existing prompt will be overwritten, and the remaining "duplicates" will be deleted.
- Adjusted the unit tests and added new ones for various upsert scenarios.
- For locally scheduled prompts, any pending prompt reminders should be rescheduled and updated on app launch. See: [BloggingPromptCoordinator#updatePromptsIfNeeded](https://github.com/wordpress-mobile/WordPress-iOS/blob/901e6bf5ec58e043ee02e4529a88f45e2b5f61c8/WordPress/Classes/ViewRelated/Blog/Blogging%20Prompts/BloggingPromptCoordinator.swift#L82) and [PromptRemindersScheduler#processSchedule](https://github.com/wordpress-mobile/WordPress-iOS/blob/901e6bf5ec58e043ee02e4529a88f45e2b5f61c8/WordPress/Classes/Utility/Blogging%20Prompts/PromptRemindersScheduler.swift#L148)
- With this update, the `title` and `content` properties in `BloggingPrompt` are now unused. I'll make a separate PR that cleans up the property from the Core Data model to keep this PR manageable.

## To test

> **Important**
Before testing this PR, launch a previous version of the Jetpack app to locally save some prompts from the v2 API. You can do this by going to the prompts card > ellipsis menu > "View more prompts" to load v2 prompts about 10 days back.

- Launch the Jetpack app.
- Log in to a WP.com account.
- 🔎 Verify that the blogging prompt card appears on the home dashboard.
- Tap "View all responses".
- 🔎 Verify that the Reader stream shows the posts related to the prompt.
- Go back and tap "Answer prompt".
- 🔎 Verify that the draft post is opened and shows the prompt content in quotes.
- Go back, tap the ellipsis menu, and select "View more prompts".
- 🔎 Verify that the lists are loaded correctly, and no duplicated dates are shown.

#### Prompt scheduling

- Install the app store version of Jetpack.
- Schedule a prompt.
- Update the Jetpack with this version and launch the app.
- Wait for the reminder to be fired.
- 🔎 Verify that the reminder shows the contents of the new v3 prompts.

## Regression Notes
1. Potential unintended areas of impact
Prompts may be incorrectly loaded, or there could be other unexpected behavior with the Blogging Prompt features.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manually tested the changes, as described in the testing steps.

3. What automated tests I added (or what prevented me from doing so)
Adjusted the unit tests and added some new ones.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

UI Changes testing checklist:

N/A

- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] VoiceOver.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn't complete, formatting should still respect the right-to-left layout)
- [ ] iPhone and iPad. 
- [ ] Multi-tasking: Split view and Slide over. (iPad)

dvdchr committed:
Migrate blogging prompt remote v3 logic

dvdchr committed:
Fix parsing logic for blogging prompts

dvdchr committed:
Improve docs for the request date

dvdchr committed:
Adjust unit tests

dvdchr committed:
Update upsert method to be based on dates

dvdchr committed:
Add tests for upsert

dvdchr committed:
Minor update to docs

mokagio commented on WordPress/Classes/Services/BloggingPrompts/BloggingPromptRemoteObject.swift:
```diff
@@ -0,0 +1,64 @@
+/// Encapsulates a single blogging prompt object from the v3 API.
+struct BloggingPromptRemoteObject {
+    var promptID: Int
+    var text: String
+    var attribution: String
+    var date: Date
+    var answered: Bool
+    var answeredUsersCount: Int
+    var answeredUserAvatarURLs: [URL]
+    var answeredLink: URL?
+    var answeredLinkText: String
```
Do all of these properties need to be mutable? I don't see any being modified in the diff.

dvdchr replied:
Good catch! This is addressed in https://github.com/wordpress-mobile/WordPress-iOS/pull/21952/commits/2c9b009ebabcecc25a6bef3b1d49e51aa3027bd2. 🙂

mokagio commented on WordPress/Classes/Services/BloggingPrompts/BloggingPromptsService.swift:
```diff
@@ -182,10 +185,11 @@ class BloggingPromptsService {
     ///   - blog: When supplied, the service will perform blogging prompts requests for this specified blog.
     ///     Otherwise, this falls back to the default account's primary blog.
```
Nitpick: Now that we have `api` in the signature, this documentation should be updated.

If I read the code correctly, do you think this could do?

```suggestion
    ///     Otherwise, this falls back to the default account's primary blog.
    ///   - api: When supplied, the WordPressComRestApi instance to use to fetch the prompts.
    ///     Otherwise, an default or anonymous instance will be computed based on whether there is an account available.
```

dvdchr replied:
Thanks for catching that! 👍🏼

mokagio commented on WordPress/Classes/Services/BloggingPrompts/BloggingPromptsService.swift:
```diff
@@ -260,6 +271,68 @@ private extension BloggingPromptsService {
+            case .success(let responseObject):
+                do {
+                    let data = try JSONSerialization.data(withJSONObject: responseObject, options: [])
+                    let decoder = JSONDecoder()
+                    decoder.dateDecodingStrategy = JSONDecoder.DateDecodingStrategy.supportMultipleDateFormats
+                    decoder.keyDecodingStrategy = .useDefaultKeys
```
Have you considered making the decoder an instance or static property? Just to save having to init and configure it on every response?

dvdchr replied:
Good point! Addressed in https://github.com/wordpress-mobile/WordPress-iOS/pull/21952/commits/d5ad31a40bc70afe69b87d4effe1082ca4a44c8f.

mokagio commented on WordPress/WordPressTest/BloggingPromptsServiceTests.swift:
```diff
@@ -136,39 +138,172 @@ final class BloggingPromptsServiceTests: CoreDataTestCase {
+    func test_fetchPrompt_shouldDeleteExcessPromptsWithTheSameDates() throws {
+        // use actual remote object so the request can be intercepted by HTTPStubs.
+        service = BloggingPromptsService(contextManager: contextManager, blog: blog)
+        stubFetchPromptsResponse()
+
+        let expectedPromptIDs: Set<Int> = [239, 248]
+
+        // insert existing prompts.
+        let date = try XCTUnwrap(Self.dateFormatter.date(from: "2022-05-03"))
+        makeBloggingPrompt(siteID: Int32(siteID), date: date)
+        makeBloggingPrompt(siteID: Int32(siteID), date: date)
+        makeBloggingPrompt(siteID: Int32(siteID), date: date)
+        makeBloggingPrompt(siteID: Int32(siteID), date: date)
+        makeBloggingPrompt(siteID: Int32(siteID), date: date)
+        contextManager.save(contextManager.mainContext)
+
+        let expectation = expectation(description: "Fetch prompts should succeed")
+        service.fetchPrompts(from: .distantPast) { prompts in
+            // the existing prompt should have been overwritten.
+            XCTAssertEqual(prompts.count, 2)
```
Out of scope because the test was already setup this way, but as a reader with no context, it took me a while to patch together that we are doing `prompts.count == 2` as a check for the existing prompts having been deleted because in the HTTP stub we set above we read from a JSON that has two prompts. 

Did you have context getting into this test? Was that clearer for you?

dvdchr replied:
> it took me a while to patch together that we are doing prompts.count == 2 as a check for the existing prompts having been deleted because in the HTTP stub we set above we read from a JSON that has two prompts. 
> Did you have context getting into this test?

Yes, I did have context for this test because I was the one who created the test originally. However, I see your point, since the assertions are hardcoded... and it would take some extra steps to connect the dots and understand that this is closely tied to the values in the stub file.

I've tried to get rid of almost all the hardcoded values (save for a few assumptions about the test data) through https://github.com/wordpress-mobile/WordPress-iOS/pull/21952/commits/176c53f8d9fe2409ec84b528bb805556f1fb2da8. Let me know if this is more readable for you overall. 🙂

mokagio replied:
Thanks @dvdchr . Your changes are definitely an improvement. They make it clearer where the expected data comes from.

I think there's room for improvement in our HTTP stubs. Personally, I'd like to move away from JSONs, which keep the source of truth for the expected data in a different file than the test, and have something were we can set the expected values as close as possible to where we assert them. But that's a discussion for another day and a change that would be far wider in reach. 👍

mokagio commented on WordPress/WordPressTest/BloggingPromptsServiceTests.swift:
```diff
@@ -196,29 +331,55 @@ private extension BloggingPromptsServiceTests {
+        let prompt = BloggingPrompt.newObject(in: contextManager.mainContext)!
+        prompt.siteID = siteID
+        prompt.date = date
+        prompt.promptID = promptID ?? NSNumber(value: arc4random_uniform(UInt32.max)).int32Value
```
Worth noting that introducing randomness makes the tests non-deterministic. Is it worth not having to type different ids in the tests where we create more than one dummy prompt?

dvdchr replied:
Good point. I've addressed this in https://github.com/wordpress-mobile/WordPress-iOS/pull/21952/commits/841761ce677a705f17f4efd0d0ea3f026857107b.

mokagio reviewed (commented):
Thank you so much for takin the time to bring the endpoint logic over to here 🎉

dvdchr committed:
Update WordPress/Classes/Services/BloggingPrompts/BloggingPromptsService.swift

Co-authored-by: Gio Lodi <gio.lodi@automattic.com>

dvdchr committed:
Make prompt remote object properties immutable

dvdchr committed:
Move the JSONDecoder to a private static variable

dvdchr commented on WordPress/Classes/Models/BloggingPrompt+CoreDataClass.swift:
```diff
@@ -21,17 +21,17 @@ public class BloggingPrompt: NSManagedObject {
         self.text = remotePrompt.text
-        self.title = remotePrompt.title
-        self.content = remotePrompt.content
+        self.title = String() // TODO: Remove
+        self.content = String() // TODO: Remove
```
These will be removed in the next PR, along with the properties from the Core Data model.

dvdchr committed:
Minor refactor and improve unit test docs

dvdchr committed:
Refactor test code to load test data as objects

dvdchr committed:
Fix prompt scheduler tests

The new API v3 response comes in array, not dictionary. In the
scheduler tests, the test data are dynamically generated via code
and still outputs the prior v2 response. This caused a type mismatch
when the BloggingPromptService decodes this stubbed response.

This fixes the issue by updating the dynamic test data generator.

dvdchr commented:
The UI test [failed 3 times consecutively](https://buildkite.com/automattic/wordpress-ios/builds/18089#018b955d-d23f-4d54-8bfb-77a4f2dc76bc) after retrying. The failure seemed to be unrelated:

```
testActivityLogCardHeaderNavigation() in DashboardTests (3 times)

XCTAssertTrue failed - Activity Log card: "The Jetpack connection" activity not displayed. (/usr/local/var/buildkite-agent/builds/MV-MKE-X64-007/automattic/wordpress-ios/WordPress/UITests/Tests/DashboardTests.swift#CharacterRangeLen=0&EndingLineNumber=53&StartingLineNumber=53)
```

Will look into this more.

dvdchr committed:
Update blogging prompts API response for UI test

dvdchr committed:
Bump API mock URL path to v3

dvdchr committed:
Update stub request parameter and the enclosing response body

The v3 response wraps the responses directly in an array, instead
of a dictionary/object with a single key.

dvdchr commented:
I did make some changes to the stubbed API for WireMock, but it's actually unrelated... The UI test for `testActivityLogCardHeaderNavigation()` is passing for me locally, and I've retried it almost 10 times now. 🤔 

When downloading the test report, it looks like the runner failed to scroll further—or maybe there's a different bug that prevented the screen from scrolling further.

<img width="1099" alt="Screenshot 2023-11-03 at 23 52 28" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/1299411/ac57708d-5ce7-4346-a21a-c3eecc97a461">

The expected behavior is to scroll further down to this part, which has been succeeding locally:

<img src="https://github.com/wordpress-mobile/WordPress-iOS/assets/1299411/fca4c6ae-ff1c-4c82-9d67-49381dbbdea6" width=250>

---

In any case... Looks like the tests are passing and I'm pretty sure my latest commits are unrelated 😅 

![](https://media.giphy.com/media/Wp1SpsnWTPWwwXaoSV/giphy.gif)

wargcm commented on WordPress/Classes/Services/BloggingPrompts/BloggingPromptsService.swift:
```diff
@@ -260,6 +281,64 @@ private extension BloggingPromptsService {
+            }
+
+            if let fromDate {
+                // convert to yyyy-MM-dd format in UTC timezone so users would get the same prompts everywhere.
+                var dateString = Self.utcDateFormatter.string(from: fromDate)
```
Is using UTC here correct? I would think that this would cause the daily prompt to change for users at a random point in their day, depending on their timezone.

I checked our current implementation and it uses the [local time](https://github.com/wordpress-mobile/WordPressKit-iOS/blob/ae2d7a1afed338adc03adc013b4a3b223fc1d25d/WordPressKit/BloggingPromptsServiceRemote.swift#L40).

dvdchr replied:
@wargcm 

Yesterday, I thought our implementation with local time _was_ the faulty one... Because with UTC, everyone would see the same prompts everywhere regardless of location, which would maximize the number of answers submitted that day. 

Plus, I also thought about the web and _assumed_ that it also works this way, considering that the server, which runs in UTC, determines what prompts the user sees at the moment. And since we are switching to v3 anyway, I'd thought to change it to UTC, too.

> this would cause the daily prompt to change for users at a random point in their day, depending on their timezone.

But yeah... Now that I think about it, this makes more sense. Plus, my assumption about the web is wrong 😆. It looks like I'm getting the prompts based on my local time, despite the web having less emphasis on _today's_ prompt — unlike the apps.

So yeah, I've reverted the implementation in https://github.com/wordpress-mobile/WordPress-iOS/pull/21952/commits/8258b09f170f4a1be7fe990f98835ccc17d907f1. Can you take a look at this again, please? Thanks!

dvdchr committed:
Base the request on local time instead of UTC

wargcm reviewed (approved):
Everything worked correctly, LGTM! 🚀 I'm going to approve since my question doesn't relate to the code here.

There is one thing I'm still slightly confused about. Will we just always be stuck out-of-sync with the web now using this new API?

When I look at their current default prompt shown:
<img width="698" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/2454408/20ae36a0-42b0-4f8e-9e91-815f3feed6a8">

And compare it to ours:
<img width="344" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/2454408/17f6a753-d9c0-4046-8fd7-1a2d3ba6ffc1">

They don't match. I assume this is because web can show multiple prompts with this API. It just makes me wonder if we should be concerned at all about engagement:

<details><summary>Prompt List Screenshot</summary>

![Simulator Screenshot](https://github.com/wordpress-mobile/WordPress-iOS/assets/2454408/ea085eae-d6a9-4ed2-b904-fa1bd95ca79a)

</details>

To me, it feels like we're going to split which prompt people on the web answer vs users on mobile.

dvdchr commented:
Thanks @wargcm ! AFAIK [the web is still using the v2 API](https://github.com/Automattic/wp-calypso/pull/76518), which is why we are 3 days ahead of prod web in this PR.

That said, once the web PR is merged, we should ensure that the prompts we see for any day is the same on all platforms to maximize engagement.

In any case, I'll keep this open for Monday in case @mokagio has additional inputs.

dvdchr commented:
👋🏼 @mokagio, I'm gonna go ahead and merge this. But if you have follow-up comments, please feel free to let me know here (or in Slack) and I'll try to address them through a separate PR. 🙂
